yt/yt/library/profiling/solomon/proxy.cpp

```diff
@@ -0,0 +1,410 @@
+{
+    registrar.Parameter("public_component_names", &TThis::PublicComponentNames)
+        .Default();
+    registrar.Parameter("max_endpoints_per_request", &TThis::MaxEndpointsPerRequest)
+        .Default(30);
+}
+
+TFingerprint IEndpointProvider::TEndpoint::GetFarmFingerprint() const
+{
+    return FarmFingerprint(Address);
+}
+
+TSolomonProxy::TSolomonProxy(TSolomonProxyConfigPtr config, IPollerPtr poller)
+    : Config_(std::move(config))
+    , HttpClient_(CreateClient(New<TClientConfig>(), std::move(poller)))
+{ }
+
+void TSolomonProxy::RegisterEndpointProvider(const IEndpointProviderPtr& endpointProvider)
+{
+    ComponentNameToEndpointProvider_[endpointProvider->GetComponentName()] = endpointProvider;
+}
+
+void TSolomonProxy::UnregisterEndpointProvider(const IEndpointProviderPtr& endpointProvider)
+{
+    EraseOrCrash(ComponentNameToEndpointProvider_, endpointProvider->GetComponentName());
+}
+
+void TSolomonProxy::Register(const TString& prefix, const IServerPtr& server)
+{
+    Register(prefix, server->GetPathMatcher());
+}
+
+void TSolomonProxy::Register(const TString& prefix, const IRequestPathMatcherPtr& handlers)
+{
+    handlers->Add(prefix + "/sensors", BIND(&TSolomonProxy::HandleSensors, MakeStrong(this)));
+    // TODO(achulkov2): Expose available tags/targets?
+}
+
+void TSolomonProxy::ValidateShardingParameters(int shard_index, int shard_count)
+{
+    if (!(0 <= shard_index && shard_index < shard_count)) {
+        THROW_ERROR_EXCEPTION(
+            "Invalid sharding configuration, %v=%v must be in range [0, %v=%v)",
+            ShardIndexParameterName,
+            shard_index,
+            ShardCountParameterName,
+            shard_count);
+    }
+}
+
+THeadersPtr TSolomonProxy::PreparePullHeaders(const THeadersPtr& reqHeaders, const TOutputEncodingContext& outputContext)
+{
+    auto pullHeaders = New<THeaders>();
+    // We use this format since it is the only one that supports compression.
+    pullHeaders->Set("Accept", TString(::NMonitoring::ContentTypeByFormat(::NMonitoring::EFormat::SPACK)));
+    pullHeaders->Set("Accept-Encoding", TString(::NMonitoring::ContentEncodingByCompression(::NMonitoring::ECompression::ZSTD)));
+    // We use a separate header to propagate whether or not this was a "solomon" pull. It affects counter-to-rate conversion.
+    pullHeaders->Set(IsSolomonFormatHeaderName, (outputContext.IsSolomonFormat ? "1" : "0"));
+
+    for (const auto& forwardHeader : ForwardHeaderWhitelist) {
+        if (auto value = reqHeaders->Find(forwardHeader)) {
+            pullHeaders->Set(forwardHeader, *value);
+        }
+    }
+
+    return pullHeaders;
+}
+
+TCgiParameters TSolomonProxy::PreparePullParameters(const TCgiParameters& params)
+{
+    auto pullParams = params;
+    for (const auto& [parameterName, _] : params) {
+        if (!ForwardParameterWhitelist.contains(parameterName)) {
+            pullParams.EraseAll(parameterName);
+        }
+    }
```

Comment:
It stemmed from the fact that the interfaces were a bit different.
`TCgiParameters` has logic around escaping/unescaping and erasing forbidden items was simpler and clearer than thinking about the correct way to copy them.
Meanwhile, `THeaders` doesn't have a cheap way to traverse the header names. Also copying only relevant headers is more efficient (and doesn't actually require the allowed items to be stored in a set).

I prefer the second approach. I looked into `TCgiParameters` a bit and I think we can just use methods of the underlying `std::multimap` to copy the parameter names and values as-is. 
Also changed the whitelists to `std::vector`, since quick lookups are no longer need here.